### PR TITLE
Add tls and http apis for post handshake

### PR DESCRIPTION
**Pull request by fAuernigg**

### CODE REVIEW

**Comment by fAuernigg** [comment]
PR #712 is required to fix the tls statemachine issue during tls v1.2 renegotiation

**Comment by cspiel1** [comment]
I think we should preserve the standard naming scheme for call back function of re/baresip:
- https://github.com/baresip/re/pull/687/commits/cd16ae12fc36d52f63bc632dbf0a1934acf6b80a#r1108715770
- https://github.com/baresip/re/pull/687/commits/cd16ae12fc36d52f63bc632dbf0a1934acf6b80a#r1108727036

File: src/tls/openssl/tls.c [hunk 1]
```diff
@@ -936,6 +954,120 @@ void tls_set_verify_client(struct tls *tls)
...
+ *
+ * @param tc     TLS connection
+ * @param depth  Max depth certificate chain accepted.
+ *               A negative depth uses default depth.
+ * @param cb     SSL verify handler. If NULL default verify handler is used.
+ */
+int tls_set_verify_client_handler(struct tls_conn *tc, int depth,
+	int (*cb) (int ok, void *d), void *d)
+{
+#if !defined(LIBRESSL_VERSION_NUMBER)
```
Comment (sreimers) [inline src/tls/openssl/tls.c:1039]: I just upgraded LibreSSL for CI tests, can you try to enable these parts. At first quick look all required APIs are available now.

**Comment by cspiel1** [comment]
> Is TLS 1.2 renegotiation support needed? Maybe better only support secure TLS 1.3 handling for this use case?

Indeed, we need it for HTTPS clients against MS Azure. And they are very lazy with security updates.

**Comment by fAuernigg** [comment]
> Is TLS 1.2 renegotiation support needed? Maybe better only support secure TLS 1.3 handling for this use case?

As far as I known renegotiation in TLSv1.2 has only security flaws, if it is initiated by the client. In the test added in this PR the server initiates renegotiation.

**Comment by sreimers** [comment]
https://github.com/curl/curl/issues/3258#issuecomment-437596053

> Note renegotiation is an extremely dangerous protocol feature. It is disabled by default in BoringSSL for security reasons and was removed in TLS 1.3.

> Yeah, I wrote that text in BoringSSL. :-P Those limitations do not avoid renegotiation's problems. They are merely the bare minimum of simplifications we've been able to manage. Mid-renegotiation a lot of state is still pretty fuzzy, and renegotiation still exposes a lot of I/O patterns applications don't expect. I've rarely seen OpenSSL-consuming code that actually gets it right. And one still must be concerned with the renegotiation attacks from way back (lack of binding between the two handshakes without RI extension), as well as 3-SHAKE.

I wonder if this opens client side attacks:

https://github.com/baresip/re/pull/712/files#diff-2907ed5afee94d2389faf81e1d83c4110381960c319b5fb5d0a009cf2b170dc8R233

Since OpenSSL does not prevent this.

**Comment by fAuernigg** [comment]
I assume our only chance to get #712 to main branch is by making it configurable like in boring ssl?
Would this be acceptable?

A http client could then enable it only for certain http connections.

**Comment by fAuernigg** [comment]
> I assume our only chance to get #712 to main branch is by making it configurable like in boring ssl? Would this be acceptable?
> 
> A http client could then enable it only for certain http connections.

TLS v1.2 renegotiation was removed from this PR. (Preparing another PR with TLSv1.2 enable api per connection)

File: src/http/server.c [hunk 2]
```diff
@@ -29,6 +30,7 @@ struct http_sock {
 	struct tcp_sock *ts;
 	struct tls *tls;
 	http_req_h *reqh;
+	https_verify_msg_h *verify_msg_h;
```
Comment (cspiel1) [inline src/http/server.c:33]: rename to:
`verify_msg_h` --> `verifyh`

File: src/http/server.c [hunk 3]
```diff
@@ -363,6 +466,33 @@ int https_listen(struct http_sock **sockp, const struct sa *laddr,
...
+ *
+ * @param sock 		HTTP socket
+ * @param verify_msg_h 	Verify handler called before the https request
+ *                      handler is called to return a http response
+ *                      to the client.
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int  https_set_verify_msgh(struct http_sock *sock,
+	https_verify_msg_h *verify_msg_h)
```
Comment (cspiel1) [inline src/http/server.c:485]: rename
`verify_msg_h` --> `verifyh`

File: src/http/server.c [hunk 4]
```diff
@@ -363,6 +466,33 @@ int https_listen(struct http_sock **sockp, const struct sa *laddr,
...
+{
+	if (!sock || !verify_msg_h)
+		return EINVAL;
+
+	sock->verify_msg_h = verify_msg_h;
```
Comment (cspiel1) [inline src/http/server.c:490]: rename to
`sock->verifyh = verifyh;`

File: include/re_http.h [hunk 5]
```diff
@@ -179,19 +179,31 @@ int http_client_set_tls_max_version(struct http_cli *cli, int version);
...
+	const struct http_msg *msg, void *arg);
 
 int http_listen_fd(struct http_sock **sockp, re_sock_t fd, http_req_h *reqh,
 		   void *arg);
 int  http_listen(struct http_sock **sockp, const struct sa *laddr,
 		 http_req_h *reqh, void *arg);
 int  https_listen(struct http_sock **sockp, const struct sa *laddr,
 		  const char *cert, http_req_h *reqh, void *arg);
+int  https_set_verify_msgh(struct http_sock *sock,
+			   https_verify_msg_h *verify_msg_h);
```
Comment (cspiel1) [inline include/re_http.h:200]: rename to
```c
int  https_set_verify_msgh(struct http_sock *sock,
			   https_verify_msg_h *verifyh);
```

**Review by cspiel1 — CHANGES_REQUESTED** [review]
Took the naming scheme from `src/dns/client.c`.

**Comment by alfredh** [comment]
I have some more comments, please take them with a grain of salt :)

There are changes to HTTP and TLS modules, and the re-negotiation feature belongs to TLS.
There are probably other protocols sitting on top of TLS which may need the same feature,
for example SIP.

Is it possible to add the feature to TLS module so that all users of TLS gets
the feature transparently ?

For example in retest this means changes to tls.c only, not http.c

If this is possible, we could split this patch in two parts, where the first part
is for TLS and the second part is for HTTP.

File: src/http/server.c [hunk 6]
```diff
@@ -376,6 +506,19 @@ struct tcp_sock *http_sock_tcp(struct http_sock *sock)
...
 
 
+/**
+ * Get the TLS struct of an HTTP sock
+ *
+ * @param conn HTTP connection
+ *
+ * @return TLS struct
+ */
+struct tls *http_sock_tls(struct http_sock *sock)
```
Comment (alfredh) [inline src/http/server.c:516]: can the input param be const ?

File: src/tls/openssl/tls.c [hunk 7]
```diff
@@ -205,6 +207,19 @@ int tls_verify_handler(int ok, X509_STORE_CTX *ctx)
 #endif
 
 
+static int tls_verify_idx = -1;
+static once_flag oflag = ONCE_FLAG_INIT;
```
Comment (alfredh) [inline src/tls/openssl/tls.c:211]: is it possible to avoid this global data and use the TLS context instead ?

File: src/tls/openssl/tls.c [hunk 8]
```diff
@@ -1956,3 +2088,54 @@ const struct list *tls_certs(const struct tls *tls)
...
+		err = EFAULT;
+		DEBUG_WARNING("%s: error: %m, ssl_err=%d\n", __func__, err,
+			SSL_get_error(tc->ssl, ret));
+	}
+
+	if (!(ret = SSL_do_handshake(tc->ssl))) {
+		err = EIO;
+		DEBUG_WARNING("%s: error: %m, ssl_err=%d\n", __func__, err,
+			SSL_get_error(tc->ssl, ret));
+	}
```
Comment (alfredh) [inline src/tls/openssl/tls.c:2138]: global openssl error queue must be cleared in both error cases

File: src/tls/openssl/tls.c [hunk 9]
```diff
@@ -1956,3 +2088,54 @@ const struct list *tls_certs(const struct tls *tls)
...
+int tls_verify_client_post_handshake(struct tls_conn *tc)
+{
+	int ret;
+	int err = 0;
+	if (!tc || !tc->ssl)
+		return EINVAL;
+
+	if (!(ret=SSL_verify_client_post_handshake(tc->ssl))) {
+		err = EFAULT;
+		DEBUG_WARNING("%s: error: %m, ssl_err=%d\n", __func__, err,
```
Comment (alfredh) [inline src/tls/openssl/tls.c:2130]: please avoid __func__ as it makes it harder to grep the code

**Comment by fAuernigg** [comment]
> I have some more comments, please take them with a grain of salt :)
> 
> There are changes to HTTP and TLS modules, and the re-negotiation feature belongs to TLS. There are probably other protocols sitting on top of TLS which may need the same feature, for example SIP.
> 
> Is it possible to add the feature to TLS module so that all users of TLS gets the feature transparently ?
> 
> For example in retest this means changes to tls.c only, not http.c
> 
> If this is possible, we could split this patch in two parts, where the first part is for TLS and the second part is for HTTP.


I'm afraid but there is no transparent tls post handshake on server side.

For example a http server requests a post handshake (using SSL_verify_client_post_handshake)  after it received http data. (Typically a specific http path, which requires more authentication data from the client).

Clients can support tls posthandshake transparently by setting for example a client certificate before a https request and enabling post handshake support using SSL_CTX_set_post_handshake_auth.


Anyhow of course I can split up the tls and http commit to make it easier to implement posthandshake in for example sip:

A sip server side implementation would need to support what "https_set_verify_msgh" of this PR does:
- Allow to set a callback receiving OSI level 7 data (e.g. comparable to http data and http path) upon which it can decide, if a post handshake should be done.

File: src/tls/openssl/tls.c (hunk 7 above)
Comment (fAuernigg) [reply src/tls/openssl/tls.c:211]: I was trying to do that but it is not possible. It is not nice but openssl requires global indices initialized once. in tls_alloc call_once(&oflag, tls_init_verify_idx) is called.

Code examples are taken from openssl and apache2/httpd implementation to get application inside the openssl verify handlers, which are not supporting application data (so no tls context is present). So the only way to get application data is to use global openssl indices initialized once. (SSL_set_ex_data, SSL_get_ex_data(..., tls_verify_idx))

**Review by cspiel1 — APPROVED** [review]

**Review by cspiel1 — APPROVED** [review]

**Pull request merged**